### PR TITLE
edit tutorials documentation

### PR DESCRIPTION
**Pull request by maccum**
This is the second piece of #4064 that I broke off. 

* change layout of tutorials landing page so that clicking on the tutorial title in the sidebar brings you to the actual tutorial
* remove types and missingness tutorials
* add subsections to api.rst, because it is getting quite large in the sidebar. 
* a few edits to tutorial contents

### CODE REVIEW

File: python/hail/docs/api.rst [hunk 1]
```diff
@@ -9,38 +9,43 @@ on the Python programming interface.
...
 .. autosummary::
     :nosignatures:
     :toctree: ./
     :template: class.rst
 
-    hail.Table
-    hail.GroupedTable
-    hail.MatrixTable
-    hail.GroupedMatrixTable
+        hail.Table
```
Comment (tpoterba) [inline python/hail/docs/api.rst]: this seems wrong - it should be one tab over from the directive

File: python/hail/docs/api.rst [hunk 2]
```diff
@@ -9,38 +9,43 @@ on the Python programming interface.
...
-    methods <methods/index>
-    utils <utils/index>
-    linalg <linalg/index>
-    stats <stats/index>
-    genetics <genetics/index>
-    plot <plot>
-    experimental <experimental>
-
-.. rubric:: Module functions
+        expressions <expressions>
```
Comment (tpoterba) [inline python/hail/docs/api.rst]: again, why this change?

File: python/hail/docs/tutorials/01-genome-wide-association-study.ipynb [hunk 3]
```diff
@@ -103,7 +111,9 @@
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
+   "metadata": {
```
Comment (tpoterba) [inline python/hail/docs/tutorials/01-genome-wide-association-study.ipynb]: we should have a tool to strip metadata out so things stay consistent. I'll write that this afternoon.

File: python/hail/docs/tutorials/08-joins.ipynb [hunk 4]
```diff
@@ -172,11 +185,33 @@
...
+   "outputs": [],
+   "source": [
+    "t2[t1.a].show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Since we only want the field `x` from `t2`, we can select it with `t2[t1.a].x`. Then we add this field to `t1` with the `anntotate_rows()` method. The new joined table `j` has a field `t2_x` that comes from the rows of `t2`. The tables could be joined, because they shared the same number of keys (1) and the same key type (string). The keys do not need to share the same name. Notice that the key values present in `t2` but not in `t1` don't show up in the final result. This join syntax performs a left join. For other types of joins, see the [join()](https://hail.is/docs/devel/hail.Table.html?highlight=join#hail.Table.join) method.\n",
```
Comment (jigold) [inline python/hail/docs/tutorials/08-joins.ipynb]: Is the new field really called `t2_x`? I would have thought it was just `x`.

File: python/hail/docs/tutorials/08-joins.ipynb [hunk 5]
```diff
@@ -172,11 +185,33 @@
...
+    "\n",
+    "The magic of keys is that they turn tables into maps: `table[expr]` should naturally refer to the row of `table` that has a key value of `expr`. Note: if the row is not unique, one such row is chosen arbitrarily.\n",
```
Comment (jigold) [inline python/hail/docs/tutorials/08-joins.ipynb]: I think **maps** should be replaced with dictionary. It will be much clearer to a Python user what the analogy is.

File: python/hail/docs/tutorials/08-joins.ipynb [hunk 6]
```diff
@@ -322,9 +345,7 @@
     }
    },
    "source": [
-    "`Table`s also have a SQL-style inner/left/right/outer [join](https://hail.is/docs/devel/hail.Table.html#hail.Table.join) method.\n",
-    "\n",
-    "SQL-style joins for `MatrixTable` are coming soon."
+    "`Table`s also have a SQL-style inner/left/right/outer [join](https://hail.is/docs/devel/hail.Table.html#hail.Table.join) method."
```
Comment (jigold) [inline python/hail/docs/tutorials/08-joins.ipynb]: I think this was already referred to above.

File: python/hail/docs/tutorials/09-matrixtable.ipynb [hunk 7]
```diff
@@ -26,7 +26,7 @@
     }
    },
    "source": [
-    "In the last example, the ratings table had a compound key: `movie_id` and `user_id`.  The ratings were secretly a movie-by-user matrix!\n",
+    "In the last example of the Table Joins Tutorial, the ratings table had a compound key: `movie_id` and `user_id`.  The ratings were secretly a movie-by-user matrix!\n",
```
Comment (jigold) [inline python/hail/docs/tutorials/09-matrixtable.ipynb]: I'd put the link to the tutorial page.

File: python/hail/docs/tutorials/04-missingness.ipynb [hunk 8]
```diff
@@ -1,267 +0,0 @@
...
-    "slideshow": {
-     "slide_type": "slide"
-    }
-   },
-   "source": [
-    "## Missingness\n",
-    "\n",
-    "All values in Hail can be missing.\n",
-    "\n",
-    "Expressions deal with missingness in a natural way. For example:\n",
```
[the remark this answers is not in the document]
Comment (maccum) [reply python/hail/docs/tutorials/04-missingness.ipynb]: Added a section on missingness to expr tutorial.

File: python/hail/docs/tutorials/05-tables.ipynb [hunk 9]
```diff
@@ -8,11 +8,11 @@
     }
    },
    "source": [
-    "## Table\n",
+    "## Table Tutorial\n",
     "\n",
     "[Table](https://hail.is/docs/devel/hail.Table.html) is Hail's distributed analogue of a data frame or SQL table.  It will be familiar if you've used R or `pandas`, but `Table` differs in 3 important ways:\n",
     "\n",
-    "- It is distributed.  Hail `Table`s can store far more data than can fit on a single computer.\n",
+    "- It is distributed.  Hail Tables can store far more data than can fit on a single computer.\n",
```
[the remark this answers is not in the document]
Comment (maccum) [reply python/hail/docs/tutorials/05-tables.ipynb]: I agree.

File: python/hail/docs/tutorials/03-expressions.ipynb [hunk 10]
```diff
@@ -33,39 +98,65 @@
...
    "source": [
-    "1 + 2"
+    "hl.int32(3)"
    ]
   },
   {
    "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `literal` function will convert a Python object into a Hail expression. Let's try it out on a Python list."
```
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: I think we should add links to the docs for the literal, value, and show functions.

File: python/hail/docs/tutorials/03-expressions.ipynb [hunk 11]
```diff
@@ -25,6 +25,71 @@
...
+    "* container types: [arrays](https://hail.is/docs/devel/types.html#hail.expr.types.tarray), [sets](https://hail.is/docs/devel/types.html#hail.expr.types.tset), [dicts](https://hail.is/docs/devel/types.html#hail.expr.types.tdict), [tuples](https://hail.is/docs/devel/types.html#hail.expr.types.ttuple), [structs](https://hail.is/docs/devel/types.html#hail.expr.types.tstruct), [intervals](https://hail.is/docs/devel/types.html#hail.expr.types.tinterval)\n",
+    "* genetics types: [locus](https://hail.is/docs/devel/types.html#hail.expr.types.tlocus), [call](https://hail.is/docs/devel/types.html#hail.expr.types.tcall)\n",
+    "\n",
+    "Hail data types are distinct from Python data types. A Hail string has a Hail type, `str`, but its Python type is a `StringExpression`. "
```
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: I think you can remove this last sentence with the examples. I find it confusing and it's covered directly below.

File: python/hail/docs/tutorials/03-expressions.ipynb (hunk 10 above)
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: convert **any** Python object

File: python/hail/docs/tutorials/03-expressions.ipynb [hunk 12]
```diff
@@ -76,9 +167,9 @@
     }
    },
    "source": [
-    "In order to process datasets of this size, Hail uses lazy evaluation.\n",
+    "Eager evaluation won't work on datasets that won't fit in memory. Consider the UK Biobank BGEN file, which is ~2TB but decompresses to >100TB in memory.\n",
     "\n",
-    "When you enter an expression, Hail doesn't execute the expression immediately: it simply records what you asked to do."
+    "In order to process datasets of this size, Hail uses **lazy** evaluation. When you enter an expression, Hail doesn't execute the expression immediately: it simply records what you asked to do."
```
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: I think a semi-colon or two dashes is correct over a colon separating the two phrases.

File: python/hail/docs/tutorials/03-expressions.ipynb [hunk 13]
```diff
@@ -142,6 +241,91 @@
...
+    "This is different from Python's treatment of missingness, where `None + 5` would produce an error. In Hail, `hl.null(hl.tint32) + 5` produces a missing result, not an error. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "hl.is_missing(hl.null(hl.tint32)+5).value"
```
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: Can you put spaces next to the plus sign?

File: python/hail/docs/tutorials/03-expressions.ipynb [hunk 14]
```diff
@@ -142,6 +241,91 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Expressions handle missingness in the following ways:\n",
+    "\n",
+    "* a missing value plus another value is always missing\n",
+    "* a conditional statement with a missing predicate is missing\n",
+    "* when aggregating a sum of values, the missing values are ignored\n",
+    "\n",
+    "This is different from Python's treatment of missingness, where `None + 5` would produce an error. In Hail, `hl.null(hl.tint32) + 5` produces a missing result, not an error. "
```
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: Maybe point out the `is_missing` and `is_defined` functions can be used to test for missingess. Also, maybe introduce what `or_missing` does.

File: python/hail/docs/tutorials/03-expressions.ipynb [hunk 15]
```diff
@@ -142,6 +241,91 @@
...
+    "hl.or_missing(True, hl.int32(5)).value"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(hl.or_missing(False, hl.int32(5)).value)"
```
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: Does this need the print statement?

File: python/hail/docs/tutorials/03-expressions.ipynb [hunk 16]
```diff
@@ -293,9 +498,11 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# `show`, `take`, and `collect`\n",
+    "### `show`, `take`, and `collect`\n",
+    "\n",
+    "Although expressions with indices have no `value`, you can use `show` to print the first few values, or `take` and `collect` to localize values to Python. \n",
```
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: Although expressions with indices do not have a single realizable value (calling the `value` method will fail), you can ...

Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: to localize all values to Python as a list of values.

File: python/hail/docs/tutorials/08-joins.ipynb [hunk 17]
```diff
@@ -172,11 +185,33 @@
...
+    "\n",
+    "The magic of keys is that they can be used to create a mapping, like a Python dictionary, between the keys of one table and the row values of another table: `table[expr]` will refer to the row of `table` that has a key value of `expr`. If the row is not unique, one such row is chosen arbitrarily.\n",
     "\n",
-    "Also note that while they look similar, `table['field1']` and `table[table2.key]` are doing very different things!"
+    "Here's a subtle bit: if `expr` is an expression indexed by a row of `table2`, then `table[expr]` is also an expression indexed by a row of `table2`.\n",
+    "\n",
+    "Also note that while they look similar, `table['field']` and `table1[table2.key]` are doing very different things!\n",
+    "\n",
+    "`table['field']` is selected a field from the table, while `table1[table2.key]` is creating a mapping between the keys of `table2` and the rows of `table1`."
```
Comment (jigold) [inline python/hail/docs/tutorials/08-joins.ipynb]: is selecting a field from `table`

File: python/hail/docs/tutorials/08-joins.ipynb [hunk 18]
```diff
@@ -172,11 +185,33 @@
...
+   "outputs": [],
+   "source": [
+    "t2[t1.a].show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Since we only want the field `x` from `t2`, we can select it with `t2[t1.a].x`. Then we add this field to `t1` with the `anntotate_rows()` method. The new joined table `j` has a field `t2_x` that comes from the rows of `t2`. The tables could be joined, because they shared the same number of keys (1) and the same key type (string). The keys do not need to share the same name. Notice that the key values present in `t2` but not in `t1` don't show up in the final result. This join syntax performs a left join. Tables also have a SQL-style inner/left/right/outer [join()](https://hail.is/docs/devel/hail.Table.html?highlight=join#hail.Table.join) method.\n",
```
Comment (jigold) [inline python/hail/docs/tutorials/08-joins.ipynb]: I don't think the sentence `Notice... result.` is correct. Maybe you meant the non-key fields. Also I like do not rather than don't in this case.

File: python/hail/docs/tutorials/03-expressions.ipynb (hunk 15 above)
Comment (maccum) [reply python/hail/docs/tutorials/03-expressions.ipynb]: unfortunately, it won't print `None` unless the print is there.

File: python/hail/docs/tutorials/03-expressions.ipynb [hunk 19]
```diff
@@ -8,7 +8,13 @@
     }
    },
    "source": [
-    "## Expression Tutorial"
+    "## Expression Tutorial\n",
+    "\n",
+    "This tutorial covers data representation with Hail's expression classes. We will go over Hail's data types and the expressions that represent them, as well as a few features of expressions, such as lazy evaluation and missingness. We will also cover how expressions can refer to fields in a table or matrix table.\n",
+    "\n",
```
Comment (jigold) [inline python/hail/docs/tutorials/03-expressions.ipynb]: Awesome!
